### PR TITLE
[MERGE] mass_mailing(_sms): move from onchange / default to stored editable computed fields

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Try to move from onchange / default_get to stored editable computed fields.
Normally behavior should be the same (computed or set by user), with support
in create / write + onchange without additional code.

SPECIFICATIONS

Update classic fields with onchange to stored editable computed fields. It
means their value will come either from manual user input, either computed
based on triggers. Purpose is to remove all onchange and default_get when
possible.

Clean fields definition inconsistencies, like default / required on computed
fields. Indeed computed fields should always have a value, maybe coming from
user input. They should not have default / required that are attributes for
classic fields.

Mailing domain computation is also cleaned and is now reset once the
mailing domain is changed. Indeed trying to keep a previous domain
based on try / except has no meaning from functional point of view. If model
changes then domain should change as it is a business record, not just a
technical domain to try to apply.

Finally add some tests to help develop and maintain those fields computation.

LINKS

Task ID 2088577
PR #41877
Enterprise PR odoo/enterprise#7278

### CODE REVIEW

File: addons/mass_mailing_event/models/event.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def action_mass_mailing_attendees(self):
             target='current',
             context=dict(
                 default_mailing_model_id=self.env.ref('event.model_event_registration').id,
-                default_mailing_domain=domain,
+                default_mailing_domain=repr(domain),
```
Comment (awa-odoo) [inline addons/mass_mailing_event/models/event.py]: `repr` on a string?

File: addons/mass_mailing_event/models/mailing_mailing.py [hunk 2]
```diff
@@ -1,14 +0,0 @@
...
-    _inherit = "mailing.mailing"
-
-    @api.depends('mailing_model_real', 'contact_list_ids')
-    def _compute_mailing_domain(self):
-        # TDE FIXME: whuuut ?
-        result = super(MassMailingCampaign, self)._compute_mailing_domain()
-        for mailing in self:
-            if mailing.mailing_model_name == 'event.registration' and mailing.mailing_domain == '[]':
-                mailing.mailing_domain = self.env.context.get('default_mailing_domain', '[]')
-        return result
```
Comment (awa-odoo) [inline addons/mass_mailing_event/models/mailing_mailing.py]: Nice FIXME :D

File: addons/mass_mailing_event_track/models/event.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def action_mass_mailing_track_speakers(self):
             target='current',
             context=dict(
                 default_mailing_model_id=self.env.ref('website_event_track.model_event_track').id,
-                default_mailing_domain="[('event_id', 'in', %s), ('stage_id.is_cancel', '!=', True)]" % self.ids,
+                default_mailing_domain=repr([('event_id', 'in', self.ids), ('stage_id.is_cancel', '!=', True)]),
```
Comment (awa-odoo) [inline addons/mass_mailing_event_track/models/event.py]: "repr" will evaluate to a string containing "self.ids" right?
I think we specifically want `"[('event_id', 'in', [1,2,3])]"` here and not `"[('event_id', 'in', self.ids)]"`.

Eeeeeeeh not sure 🤷‍♂️

File: addons/mass_mailing_slides/models/slide_channel.py [hunk 4]
```diff
@@ -8,7 +8,7 @@ class Course(models.Model):
     _inherit = "slide.channel"
 
     def action_mass_mailing_attendees(self):
-        domain = [('slide_channel_ids', 'in', self.ids)]
+        domain = repr([('slide_channel_ids', 'in', self.ids)])
```
Comment (awa-odoo) [inline addons/mass_mailing_slides/models/slide_channel.py]: Same remark here.

File: addons/mass_mailing_sms/models/mailing_mailing.py [hunk 5]
```diff
@@ -209,3 +210,14 @@ def action_send_sms(self, res_ids=None):
...
         return True
+
+    # --------------------------------------------------
+    # TOOLS
+    # --------------------------------------------------
+
+    def _get_default_mailing_domain(self):
+        mailing_domain = super(Mailing, self)._get_default_mailing_domain()
+        if self.mailing_type == 'sms' and 'phone_blacklisted' in self.env[self.mailing_model_name]._fields:
+            mailing_domain = expression.AND([mailing_domain or [], [('phone_blacklisted', '=', False)]])
```
Comment (awa-odoo) [inline addons/mass_mailing_sms/models/mailing_mailing.py]: is the ` or []` necessary?

File: addons/mass_mailing_event_track/models/event.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing_event_track/models/event.py]: Python will first evaluate the list making it [('event_id', 'in', [1,2,3])] then convert the list to its representation making it "[('event_id', 'in', [1,2,3])]".

Aka
```
a = [2, 3, 4]
repr([('id', 'in', a)])
# return "[('id', 'in', [2, 3, 4])]"
```

File: addons/mass_mailing_event/models/event.py (hunk 1 above)
Comment (awa-odoo) [reply addons/mass_mailing_event/models/event.py]: Then, based on the behavior of repr, could be `repr([('event_id', 'in', self.ids)])` all the time

Comment (tde-banana-odoo) [reply addons/mass_mailing_event/models/event.py]: Did not see it was a string. Updated.

File: addons/mass_mailing_sms/models/mailing_mailing.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing_sms/models/mailing_mailing.py]: Nope.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**